File: clients/sdk/frontend/typescript/src/types/queries.ts
// Query parameter types for frontend operations

export type ThreadListQueryType = {
  limit?: number; // 1-100 (service enforced)
  before?: string; // pagination anchor for newer threads
  after?: string;  // pagination anchor for older threads
  anchor?: string;  // specific pagination anchor
  sort_by?: 'created_ts' | 'updated_ts'; // sort field
};

export type MessageListQueryType = {
  limit?: number; // 1-100 (service enforced)
  before?: string; // pagination anchor for older messages
  after?: string;  // pagination anchor for newer messages
  anchor?: string;  // specific pagination anchor
  sort_by?: 'created_ts' | 'updated_ts'; // sort field
};

File: clients/sdk/backend/python/src/progressdb/types/queries.py
from typing import TypedDict, Optional


# Base query parameters (non-pagination)
class BaseQueryType(TypedDict, total=False):
    limit: Optional[int]  # 1-100 (service enforced)
    sort_by: Optional[str]  # 'created_ts' | 'updated_ts'

# Pagination query types - mutually exclusive
class BeforeQueryType(BaseQueryType):
    before: str
    # after and anchor are excluded by mutual exclusivity

class AfterQueryType(BaseQueryType):
    after: str
    # before and anchor are excluded by mutual exclusivity

class AnchorQueryType(BaseQueryType):
    anchor: str
    # before and after are excluded by mutual exclusivity

class NoPaginationQueryType(BaseQueryType):
    # before, after, anchor are excluded by mutual exclusivity
    pass

# Combined query types with mutual exclusivity enforced
ThreadListQueryType = BeforeQueryType | AfterQueryType | AnchorQueryType | NoPaginationQueryType
MessageListQueryType = BeforeQueryType | AfterQueryType | AnchorQueryType | NoPaginationQueryType